## Expose negotiated ECDHE curve as $ssl_ecdhe_curve
## (hopefully for TLSv1.3 as well)
--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -8,6 +8,7 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_event.h>
+#include <openssl/objects.h>
 
 
 #define NGX_SSL_PASSWORD_BUFFER_SIZE  4096
@@ -2495,6 +2496,7 @@ ngx_ssl_new_session(ngx_ssl_conn_t *ssl_
     ngx_ssl_sess_id_t        *sess_id;
     ngx_ssl_session_cache_t  *cache;
     u_char                    buf[NGX_SSL_MAX_SESSION_SIZE];
+    const char               *ciph;
 
     len = i2d_SSL_SESSION(sess, NULL);
 
@@ -2509,6 +2511,12 @@ ngx_ssl_new_session(ngx_ssl_conn_t *ssl_
 
     c = ngx_ssl_get_connection(ssl_conn);
 
+    ciph = SSL_get_cipher_name(ssl_conn);
+    if (ciph && (!strncmp("ECDHE-", ciph, 6) || !strncmp("TLS13-", ciph, 6)))
+        c->ssl->curve_nid = SSL_get_shared_curve(c->ssl->connection, 0);
+    else
+        c->ssl->curve_nid = NID_undef;
+
     ssl_ctx = c->ssl->session_ctx;
     shm_zone = SSL_CTX_get_ex_data(ssl_ctx, ngx_ssl_session_cache_index);
 
@@ -2600,6 +2608,7 @@ ngx_ssl_new_session(ngx_ssl_conn_t *ssl_
     sess_id->id = id;
     sess_id->len = len;
     sess_id->session = cached_sess;
+    sess_id->curve_nid = c->ssl->curve_nid;
 
     sess_id->expire = ngx_time() + SSL_CTX_get_timeout(ssl_ctx);
 
@@ -2697,6 +2706,7 @@ ngx_ssl_get_cached_session(ngx_ssl_conn_
 
             if (sess_id->expire > ngx_time()) {
                 ngx_memcpy(buf, sess_id->session, sess_id->len);
+                c->ssl->curve_nid = sess_id->curve_nid;
 
                 ngx_shmtx_unlock(&shpool->mutex);
 
@@ -3536,6 +3546,14 @@ ngx_ssl_get_curves(ngx_connection_t *c, 
 
 
 ngx_int_t
+ngx_ssl_get_ecdhe_curve_name(ngx_connection_t *c, ngx_pool_t *pool, ngx_str_t *s)
+{
+    s->data = (u_char *) OBJ_nid2sn(c->ssl->curve_nid);
+    return NGX_OK;
+}
+
+
+ngx_int_t
 ngx_ssl_get_session_id(ngx_connection_t *c, ngx_pool_t *pool, ngx_str_t *s)
 {
     u_char        *buf;
--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -83,6 +83,7 @@ struct ngx_ssl_connection_s {
     ngx_int_t                   last;
     ngx_buf_t                  *buf;
     size_t                      buffer_size;
+    int                         curve_nid;
 
     ngx_connection_handler_pt   handler;
 
@@ -119,6 +120,7 @@ struct ngx_ssl_sess_id_s {
     u_char                     *session;
     ngx_queue_t                 queue;
     time_t                      expire;
+    int                         curve_nid;
 #if (NGX_PTR_SIZE == 8)
     void                       *stub;
     u_char                      sess_id[32];
@@ -216,6 +218,8 @@ ngx_int_t ngx_ssl_get_ciphers(ngx_connec
     ngx_str_t *s);
 ngx_int_t ngx_ssl_get_curves(ngx_connection_t *c, ngx_pool_t *pool,
     ngx_str_t *s);
+ngx_int_t ngx_ssl_get_ecdhe_curve_name(ngx_connection_t *c, ngx_pool_t *pool,
+    ngx_str_t *s);
 ngx_int_t ngx_ssl_get_session_id(ngx_connection_t *c, ngx_pool_t *pool,
     ngx_str_t *s);
 ngx_int_t ngx_ssl_get_session_reused(ngx_connection_t *c, ngx_pool_t *pool,
--- a/src/http/modules/ngx_http_ssl_module.c
+++ b/src/http/modules/ngx_http_ssl_module.c
@@ -318,6 +318,9 @@ static ngx_http_variable_t  ngx_http_ssl
     { ngx_string("ssl_curves"), NULL, ngx_http_ssl_variable,
       (uintptr_t) ngx_ssl_get_curves, NGX_HTTP_VAR_CHANGEABLE, 0 },
 
+    { ngx_string("ssl_ecdhe_curve"), NULL, ngx_http_ssl_static_variable,
+      (uintptr_t) ngx_ssl_get_ecdhe_curve_name, NGX_HTTP_VAR_CHANGEABLE, 0 },
+
     { ngx_string("ssl_session_id"), NULL, ngx_http_ssl_variable,
       (uintptr_t) ngx_ssl_get_session_id, NGX_HTTP_VAR_CHANGEABLE, 0 },
 
--- a/src/stream/ngx_stream_ssl_module.c
+++ b/src/stream/ngx_stream_ssl_module.c
@@ -233,6 +233,9 @@ static ngx_stream_variable_t  ngx_stream
     { ngx_string("ssl_curves"), NULL, ngx_stream_ssl_variable,
       (uintptr_t) ngx_ssl_get_curves, NGX_STREAM_VAR_CHANGEABLE, 0 },
 
+    { ngx_string("ssl_ecdhe_curve"), NULL, ngx_stream_ssl_static_variable,
+      (uintptr_t) ngx_ssl_get_ecdhe_curve_name, NGX_STREAM_VAR_CHANGEABLE, 0 },
+
     { ngx_string("ssl_session_id"), NULL, ngx_stream_ssl_variable,
       (uintptr_t) ngx_ssl_get_session_id, NGX_STREAM_VAR_CHANGEABLE, 0 },
 
